### PR TITLE
Ledger restore for new architecture + more

### PR DESCRIPTION
**Pull request by snopoke**
http://manage.dimagi.com/default.asp?225617
@dimagi/scale-team 

Bit of an oversight

:fish:

### CODE REVIEW

File: corehq/apps/commtrack/tests/test_xml.py [hunk 1]
```diff
@@ -46,50 +48,58 @@ def setUp(self):
...
         user = self.user
         self.assertFalse(get_ota_balance_xml(self.domain, user))
 
+    @run_with_all_backends
     def test_ota_basic(self):
         user = self.user
-        amounts = [(p._id, i*10) for i, p in enumerate(self.products)]
-        report = _report_soh(amounts, self.sp.case_id, 'stock')
+        amounts = [
+            SohReport(section_id='stock', product_id=p._id, amount=i*10)
```
Comment (dimagimon) [inline corehq/apps/commtrack/tests/test_xml.py]: E226 missing whitespace around arithmetic operator

File: corehq/form_processor/interfaces/dbaccessors.py [hunk 2]
```diff
@@ -350,6 +350,24 @@ def get_ledger_transactions_for_case(case_id, section_id=None, entry_id=None):
 
     @abstractmethod
     def get_latest_transaction(case_id, section_id, entry_id):
+        raise NotImplementedError\
+
+    @abstractmethod
```
Comment (dimagimon) [inline corehq/form_processor/interfaces/dbaccessors.py]: E301 expected 1 blank line, found 0

File: corehq/ex-submodules/casexml/apps/phone/data_providers/case/stock.py [hunk 3]
```diff
@@ -1,6 +1,10 @@
 from collections import defaultdict
-from casexml.apps.stock.consumption import compute_consumption_or_default
+from casexml.apps.stock.consumption import compute_consumption_or_default, compute_default_monthly_consumption, \
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/data_providers/case/stock.py]: F401 'compute_consumption_or_default' imported but unused

File: corehq/form_processor/migrations/0055_daily_consumption.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+            name='daily_consumption',
+            field=models.DecimalField(null=True, max_digits=20, decimal_places=5),
+            preserve_default=True,
+        ),
+        migrations.AlterField(
+            model_name='casetransaction',
+            name='type',
+            field=models.PositiveSmallIntegerField(choices=[(1, b'form'), (2, b'rebuild_with_reason'), (4, b'user_requested_rebuild'), (8, b'user_archived_rebuild'), (16, b'form_archive_rebuild'), (32, b'form_edit_rebuild'), (64, b'ledger'), (128, b'case_create'), (256, b'case_close'), (1024, b'case_attachment'), (512, b'case_index')]),
+            preserve_default=True,
+        ),
```
Comment (snopoke) [inline corehq/form_processor/migrations/0055_daily_consumption.py]: This isn't part of this change but is a noop

**Comment by snopoke** [comment]
:white_check_mark:

**Comment by snopoke** [comment]
@benrudolph @dannyroberts

**Comment by snopoke** [comment]
cc @esoergel

**Comment by dannyroberts** [comment]
A bit hard to follow, but it looks thoroughly tested!

File: corehq/form_processor/backends/sql/ledger.py [hunk 5]
```diff
@@ -77,6 +77,7 @@ def _lazy_original_balance():
         ledger_value.track_create(transaction)
         # only do this after we've created the transaction otherwise we'll get the wrong delta
         ledger_value.balance = new_ledger_values.balance
+        ledger_value.last_modified = stock_trans.timestamp
```
Comment (dannyroberts) [inline corehq/form_processor/backends/sql/ledger.py]: Does this mean that `ledger_value.last_modified` is changing from representing approximate submit time (off by a few msec-to-sec) to representing completion time? (And if so, just checking that's desired.)

File: corehq/form_processor/models.py [hunk 6]
```diff
@@ -1216,6 +1216,14 @@ class LedgerValue(DisabledDbMixin, models.Model, TrackRelatedChanges):
     last_modified = models.DateTimeField(auto_now=True)
 
     @property
+    def last_modified_date(self):
+        return self.last_modified
+
+    @property
+    def product_id(self):
+        return self.entry_id
```
Comment (dannyroberts) [inline corehq/form_processor/models.py]: Nice

Comment (dannyroberts) [inline corehq/form_processor/models.py]: Although also pretty confusing if the code sometimes calls it one and sometimes calls it the other—someone seeing that would assume there's some subtle difference

**Comment by dannyroberts** [comment]
Just took a less distracted look today and made a few comments, but looks really good

File: corehq/form_processor/models.py (hunk 6 above)
Comment (snopoke) [inline corehq/form_processor/models.py]: Yea, this is basically just for compatibility with the `StockState` class so that they can be used interchangeably during restore (and other places) without additional boilerplate to check which property name is needed.

Comment (dannyroberts) [inline corehq/form_processor/models.py]: 👍 figured. Sounds like a mildly annoying state that's better than the alternative :)

File: corehq/form_processor/backends/sql/ledger.py (hunk 5 above)
Comment (snopoke) [inline corehq/form_processor/backends/sql/ledger.py]: Good question. `stock_trans.timestamp` is actually the date from the stock report entry in the XML:
```xml
<balance date="date from form" entity-id="..." section-id="...">
	....
</balance>
```
This change copies the current functionality: https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/casexml/apps/stock/signals.py#L113

I think the date is only used to show in reports and also in generating the restore XML: https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/casexml/apps/phone/data_providers/case/stock.py#L42

To be honest I would prefer this to match up with `form.received_on` but I'm not sure if that would break anything on mobile (though I don't expect it would)

Comment (czue) [inline corehq/form_processor/backends/sql/ledger.py]: It's been a long-backlogged item to convert to using server times for ledger dates and order logic. If it's an easy change to make now we should do it - I don't think it will break anything.

Comment (snopoke) [inline corehq/form_processor/backends/sql/ledger.py]: https://github.com/dimagi/commcare-hq/pull/11508

Comment (esoergel) [inline corehq/form_processor/backends/sql/ledger.py]: One downside with using server times for ledger dates is that users can only submit stock updates in real-time.  For instance, I worked on a project where the mobile workers would go around at the end of each month and record the EOD totals for each day.

Comment (snopoke) [inline corehq/form_processor/backends/sql/ledger.py]: These dates don't seem to be used for anything though (see #11508).
